File: backend/physics/collisions.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Optional
import torch
from .particle import ParticleSystem


@dataclass
class Bounds:
    """Axis-aligned bounding box."""
    min_pos: torch.Tensor  # (dim,) float32
    max_pos: torch.Tensor  # (dim,) float32


def enforce_bounds_clamp(ps: ParticleSystem, bounds: Bounds) -> None:
    """Clamp particles to bounding box. Ported from traerphysics.js enforceBounds (lines 258-265)."""
    ps.positions = torch.clamp(ps.positions, min=bounds.min_pos, max=bounds.max_pos)


def enforce_bounds_elastic(ps: ParticleSystem, bounds: Bounds,
                           radii: Optional[torch.Tensor] = None,
                           restitution: float = 0.8) -> None:
    """Elastic boundary collision with velocity reflection.
    Ported from mesh.html Particle.checkEdges (lines 64-87).
    restitution: coefficient of restitution (0=fully inelastic, 1=perfect bounce)."""
    r = radii if radii is not None else torch.zeros(ps.count)

    for d in range(ps.dim):
        low = bounds.min_pos[d]
        high = bounds.max_pos[d]

        below = ps.positions[:, d] - r < low
        above = ps.positions[:, d] + r > high

        ps.positions[below, d] = low + r[below]
        ps.velocities[below, d] *= -restitution

        ps.positions[above, d] = high - r[above]
        ps.velocities[above, d] *= -restitution


def floor_collision(ps: ParticleSystem, floor_y: float, restitution: float = 0.3) -> None:
    """Floor collision with bounce. Ported from traer.py (lines 94-97).
    Uses the Y axis (index 1) as vertical."""
    below = ps.positions[:, 1] < floor_y
    ps.positions[below, 1] = floor_y
    ps.velocities[below, 1] *= -restitution


def resolve_elastic_collisions(ps: ParticleSystem, radii: torch.Tensor) -> None:
    """Pairwise elastic particle collisions with impulse response.
    Ported from mesh.html Particle.checkCollision (lines 89-107).
    Brute-force O(N^2) — suitable for small particle counts."""
    n = ps.count
    if n < 2:
        return

    # Pairwise distances
    dists = torch.cdist(ps.positions, ps.positions)  # (N, N)
    min_dists = radii.unsqueeze(0) + radii.unsqueeze(1)  # (N, N)

    # Find colliding pairs (upper triangle only to avoid double-processing)
    mask = (dists < min_dists) & (dists > 1e-6)
    mask = torch.triu(mask, diagonal=1)

    pairs = mask.nonzero()  # (P, 2)
    if pairs.shape[0] == 0:
        return

    i = pairs[:, 0]
    j = pairs[:, 1]

    pos_i = ps.positions[i]
    pos_j = ps.positions[j]
    vel_i = ps.velocities[i]
    vel_j = ps.velocities[j]
    mass_i = ps.masses[i]
    mass_j = ps.masses[j]

    # Normal direction
    delta = pos_j - pos_i
    dist = delta.norm(dim=-1, keepdim=True).clamp(min=1e-6)
    normal = delta / dist

    # Relative velocity
    rel_vel = vel_j - vel_i
    vel_along_normal = (rel_vel * normal).sum(dim=-1, keepdim=True)

    # Impulse (equal mass simplification: impulse = vel_along_normal)
    total_mass = mass_i.unsqueeze(-1) + mass_j.unsqueeze(-1)
    impulse = 2.0 * vel_along_normal * normal / total_mass

    # Apply impulse
    impulse_i = impulse * mass_j.unsqueeze(-1)
    impulse_j = impulse * mass_i.unsqueeze(-1)

    # Position separation
    overlap = (radii[i] + radii[j]).unsqueeze(-1) - dist
    separation = normal * overlap * 0.5

    # Scatter updates
    ps.velocities.index_add_(0, i, impulse_i)
    ps.velocities.index_add_(0, j, -impulse_j)
    ps.positions.index_add_(0, i, -separation)
    ps.positions.index_add_(0, j, separation)

File: frontend/src/physics/collisions.ts
import type { ParticleSystem } from "./ParticleSystem";

export interface Bounds {
  minPos: Float32Array;
  maxPos: Float32Array;
}

export function enforceBoundsClamp(ps: ParticleSystem, bounds: Bounds): void {
  const dim = ps.dim;
  const pos = ps.positions;
  for (let i = 0; i < ps.count; i++) {
    for (let d = 0; d < dim; d++) {
      const idx = i * dim + d;
      if (pos[idx] < bounds.minPos[d]) pos[idx] = bounds.minPos[d];
      if (pos[idx] > bounds.maxPos[d]) pos[idx] = bounds.maxPos[d];
    }
  }
}

export function enforceBoundsElastic(
  ps: ParticleSystem,
  bounds: Bounds,
  radii: Float32Array | null = null,
  restitution: number = 0.8,
): void {
  const dim = ps.dim;
  const pos = ps.positions;
  const vel = ps.velocities;

  for (let i = 0; i < ps.count; i++) {
    const r = radii ? radii[i] : 0;
    for (let d = 0; d < dim; d++) {
      const idx = i * dim + d;
      if (pos[idx] - r < bounds.minPos[d]) {
        pos[idx] = bounds.minPos[d] + r;
        vel[idx] *= -restitution;
      }
      if (pos[idx] + r > bounds.maxPos[d]) {
        pos[idx] = bounds.maxPos[d] - r;
        vel[idx] *= -restitution;
      }
    }
  }
}

export function floorCollision(
  ps: ParticleSystem,
  floorY: number,
  restitution: number = 0.3,
): void {
  const dim = ps.dim;
  const pos = ps.positions;
  const vel = ps.velocities;

  for (let i = 0; i < ps.count; i++) {
    const yIdx = i * dim + 1;
    if (pos[yIdx] < floorY) {
      pos[yIdx] = floorY;
      vel[yIdx] *= -restitution;
    }
  }
}

export function resolveElasticCollisions(
  ps: ParticleSystem,
  radii: Float32Array,
): void {
  const n = ps.count;
  if (n < 2) return;
  const dim = ps.dim;
  const pos = ps.positions;
  const vel = ps.velocities;
  const masses = ps.masses;

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      // Distance between particles
      let distSq = 0;
      for (let d = 0; d < dim; d++) {
        const dd = pos[j * dim + d] - pos[i * dim + d];
        distSq += dd * dd;
      }
      const dist = Math.sqrt(distSq);
      const minDist = radii[i] + radii[j];

      if (dist >= minDist || dist < 1e-6) continue;

      const invDist = 1.0 / dist;

      // Relative velocity along normal
      let velAlongNormal = 0;
      for (let d = 0; d < dim; d++) {
        const normal = (pos[j * dim + d] - pos[i * dim + d]) * invDist;
        velAlongNormal += (vel[j * dim + d] - vel[i * dim + d]) * normal;
      }

      // Impulse
      const totalMass = masses[i] + masses[j];
      const impulseMag = (2.0 * velAlongNormal) / totalMass;

      // Overlap separation
      const overlap = minDist - dist;

      for (let d = 0; d < dim; d++) {
        const normal = (pos[j * dim + d] - pos[i * dim + d]) * invDist;
        const impulse = impulseMag * normal;

        vel[i * dim + d] += impulse * masses[j];
        vel[j * dim + d] -= impulse * masses[i];

        const sep = normal * overlap * 0.5;
        pos[i * dim + d] -= sep;
        pos[j * dim + d] += sep;
      }
    }
  }
}
